PSR-16 compatible cache interface

burzum opened the pull request.

burzum committed:
Started to work on the PSR6 / PSR16 implementation

burzum committed:
Made cache tests run without fatal errors again

burzum committed:
Adding writeMany() back

burzum committed:
Updating the RedisEngine

burzum committed:
Refactoring more cache engines for PSR6

burzum committed:
Fixing the Cache tests

stickler-ci committed:
Fixing style errors.

ADmad commented on src/Cache/Engine/ApcuEngine.php:
```diff
@@ -68,15 +75,21 @@ public function write($key, $value)
 
-        return apcu_fetch($key);
+        $result = apcu_fetch($key);
+        if ($result === false) {
```
I don't think this is correct. The success or failure of `apcu_fetch()` is determined by passing 2nd argument to the method.

http://php.net/manual/en/function.apcu-fetch.php

ADmad commented on src/Cache/Engine/WincacheEngine.php:
```diff
@@ -70,14 +73,20 @@ public function write($key, $value)
 
-        return wincache_ucache_get($key);
+        $result = wincache_ucache_get($key);
+        if ($result === false) {
```
Here too cache hit/miss is checked by passing second param to `wincache_ucache_get()`.
http://php.net/manual/en/function.wincache-ucache-get.php

burzum committed:
Adding CacheException and working on the Cache API

burzum committed:
Merge branch '3.next-cache' of github.com:burzum/cakephp into 3.next-cache

# Conflicts:
#	src/Cache/Cache.php
#	src/Cache/CacheEngine.php

stickler-ci committed:
Fixing style errors.

dereuromark commented on src/Cache/Engine/RedisEngine.php:
```diff
@@ -121,47 +119,42 @@ protected function _connect()
+     * @inheritDoc
      */
-    public function read($key)
+    public function get($key, $default = null)
```
If it is not a partial one, the {} are wrong.

burzum committed:
Fixing apcu and wincache engines get()

burzum committed:
Fixing Cache::read()

Needs to call the deprecated read() method on the engines, because they take care of returning this in a BC fashion

markstory commented on src/Cache/BackwardCompatibilityTrait.php:
```diff
@@ -0,0 +1,119 @@
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         1.2.0
```
This will need to be updated ⛳️

markstory commented on src/Cache/Cache.php:
```diff
@@ -362,6 +383,37 @@ public static function read($key, $config = 'default')
+     * @throws \Psr\SimpleCache\InvalidArgumentException
+     *   MUST be thrown if the $key string is not a legal value.
+     */
+    public static function get($key, $default = null, $config = 'default')
```
For the new cache interface, should we consider making the engines implement PSR16, and `Cache` becoming more of a registry/factory/container for cache engines? Right now we can't easily inject caching objects into other things. Would it be 'better' if we used this new interface implementation as an opportunity to make the interface something like

```php
// Get an engine
$cache = Cache::engine('default');
// Use the engine
$cache->set($key, $value, $ttl);
$value = $cache->get($key, $default);
```

I recognize this would be a big departure from what we have today. I think it would let us make engines more standalone and set us up for a DI container in the future.

burzum replied:
Shouldn't this be `getEngine()`? I thought to implement PSR6 first, then 16 in a follow up PR.

markstory replied:
`Cache::engine()` only functions as a getter, so it doesn't require a `get` prefix to me.

burzum replied:
I think we should use getX() it's more consistent and there is no need if we ever add a getter to refactor it.

markstory reviewed (commented):
How does PSR6 interact with our current concepts of cache groups?

markstory commented:
This is looking good @burzum. It is nice that the PSR interfaces are close to what we already have.

burzum commented:
@markstory thanks but I could need some help with the engines I have not yet installed and can't install that easy. Doing a lot of commits and waiting for Travis is pretty ugly and will take time.

markstory commented:
@burzum Which backends do you need help with?

burzum commented:
@markstory all that don't pass yet. :) Also there is one case broken that I don't know how to fix it yet. There is one test that retuns null but now false because of the BC change:

```sh
1) Cake\Test\TestCase\Cache\CacheTest::testWriteEmptyValues
Failed asserting that false is null.

tests\TestCase\Cache\CacheTest.php:628
```

burzum committed:
Minor doc block changes

ADmad commented on src/Cache/Cache.php:
```diff
@@ -240,6 +240,7 @@ public static function engine($config)
      *
      * Permanently remove all expired and deleted data
      *
+     * @deprecated Use clearExpired() instead
```
`@deprecated 3.6.0 Use....`
